File: impl/go/internal/chanlun/zhongshu.go
package chanlun

// Zhongshu (走势中枢) decomposition, faithful port of
// conformance/chanlun-v1/reference_backend/zhongshu.py:zhongshu.
//
// Scan a list of elements (each {idx, lo, hi}); a center forms iff
// max(lo of first 3) <= min(hi of first 3) (ZD <= ZG). Extend with subsequent
// elements that still overlap [zd, zg]. Zone gate:
//   "first3" - the test zone stays fixed at the first-3 overlap;
//   "all"    - re-tighten zd/zg to the running overlap of all members.
// On no-overlap, slide one element. Output is the list of
// {start, end, ZD, ZG, n} per center.

// Element is one input price-range (a stroke or segment in element form).
type Element struct {
	Idx int64
	Lo  int64
	Hi  int64
}

// Center is one emitted zhongshu.
type Center struct {
	Start int64
	End   int64
	ZD    int64
	ZG    int64
	N     int64
}

// ElementsFromValue parses a list of {idx, lo, hi}.
func ElementsFromValue(v Value) []Element {
	arr := v.MustArr()
	out := make([]Element, len(arr))
	for i, e := range arr {
		o := e.MustObj()
		idx, _ := o.Get("idx")
		lo, _ := o.Get("lo")
		hi, _ := o.Get("hi")
		out[i] = Element{Idx: idx.MustInt(), Lo: lo.MustInt(), Hi: hi.MustInt()}
	}
	return out
}

// ElementsToValue emits an element list as a Value tree.
func ElementsToValue(els []Element) Value {
	out := make([]Value, len(els))
	for i, el := range els {
		o := NewObject()
		o.Set("idx", VInt(el.Idx))
		o.Set("lo", VInt(el.Lo))
		o.Set("hi", VInt(el.Hi))
		out[i] = VObj(o)
	}
	return VArr(out)
}

// Zhongshu runs the decomposition with the named zone gate.
func Zhongshu(els []Element, zone string) []Center {
	centers := make([]Center, 0)
	n := len(els)
	i := 0
	for i+2 < n {
		ZD := maxInt(maxInt(els[i].Lo, els[i+1].Lo), els[i+2].Lo)
		ZG := minInt(minInt(els[i].Hi, els[i+1].Hi), els[i+2].Hi)
		if ZD <= ZG {
			members := []int{i, i + 1, i + 2}
			zd, zg := ZD, ZG
			j := i + 3
			for j < n && els[j].Lo <= zg && els[j].Hi >= zd {
				members = append(members, j)
				if zone == "all" {
					zd = maxInt(zd, els[j].Lo)
					zg = minInt(zg, els[j].Hi)
				}
				j++
			}
			last := members[len(members)-1]
			centers = append(centers, Center{
				Start: int64(i),
				End:   int64(last),
				ZD:    ZD,
				ZG:    ZG,
				N:     int64(len(members)),
			})
			i = last + 1
		} else {
			i++
		}
	}
	return centers
}

// CentersToValue emits a center list as a Value tree.
func CentersToValue(cs []Center) Value {
	out := make([]Value, len(cs))
	for i, c := range cs {
		o := NewObject()
		o.Set("start", VInt(c.Start))
		o.Set("end", VInt(c.End))
		o.Set("ZD", VInt(c.ZD))
		o.Set("ZG", VInt(c.ZG))
		o.Set("n", VInt(c.N))
		out[i] = VObj(o)
	}
	return VArr(out)
}

// CentersFromValue parses a list of {ZD, ZG, end, n, start} (trend_type input).
func CentersFromValue(v Value) []Center {
	arr := v.MustArr()
	out := make([]Center, len(arr))
	for i, e := range arr {
		o := e.MustObj()
		zd, _ := o.Get("ZD")
		zg, _ := o.Get("ZG")
		// trend_type fixtures may or may not carry start/end/n; default to 0.
		var start, end, n int64
		if v, ok := o.Get("start"); ok {
			start = v.MustInt()
		}
		if v, ok := o.Get("end"); ok {
			end = v.MustInt()
		}
		if v, ok := o.Get("n"); ok {
			n = v.MustInt()
		}
		out[i] = Center{Start: start, End: end, ZD: zd.MustInt(), ZG: zg.MustInt(), N: n}
	}
	return out
}

File: impl/csharp/src/Chanlun/Zhongshu.cs
// 走势中枢 decomposition — faithful port of
// conformance/chanlun-v1/reference_backend/zhongshu.py:zhongshu.
//
// Scan a list of elements (each {idx, lo, hi}); a center forms iff
// max(lo of first 3) <= min(hi of first 3) (ZD <= ZG). Extend with subsequent
// elements overlapping [zd, zg]. Zone gate:
//   "first3" - test zone stays fixed at the first-3 overlap;
//   "all"    - re-tighten zd/zg to the running overlap of all members.
// On no-overlap: slide one element. Output is the list of
// {start, end, ZD, ZG, n} per center.

namespace Chanlun;

public static class ZhongshuOps
{
    public static List<Center> Zhongshu(IReadOnlyList<Element> els, string zone)
    {
        var centers = new List<Center>();
        int n = els.Count;
        int i = 0;
        while (i + 2 < n)
        {
            long ZD = Math.Max(Math.Max(els[i].Lo, els[i + 1].Lo), els[i + 2].Lo);
            long ZG = Math.Min(Math.Min(els[i].Hi, els[i + 1].Hi), els[i + 2].Hi);
            if (ZD <= ZG)
            {
                var members = new List<int> { i, i + 1, i + 2 };
                long zd = ZD, zg = ZG;
                int j = i + 3;
                while (j < n && els[j].Lo <= zg && els[j].Hi >= zd)
                {
                    members.Add(j);
                    if (zone == "all")
                    {
                        zd = Math.Max(zd, els[j].Lo);
                        zg = Math.Min(zg, els[j].Hi);
                    }
                    j++;
                }
                int last = members[^1];
                centers.Add(new Center(i, last, ZD, ZG, members.Count));
                i = last + 1;
            }
            else
            {
                i++;
            }
        }
        return centers;
    }

    // ---- Value-tree converters ------------------------------------------------

    public static List<Element> ElementsFromValue(Value v)
    {
        var arr = v.AsArray();
        var out_ = new List<Element>(arr.Count);
        foreach (var e in arr)
        {
            var o = e.AsObject();
            out_.Add(new Element(
                o.Get("idx").AsInt(),
                o.Get("lo").AsInt(),
                o.Get("hi").AsInt()));
        }
        return out_;
    }

    public static Value ElementsToValue(IReadOnlyList<Element> els)
    {
        var arr = new List<Value>(els.Count);
        foreach (var el in els)
        {
            var o = new OrderedObject();
            o.Set("idx", Value.Of(el.Idx));
            o.Set("lo", Value.Of(el.Lo));
            o.Set("hi", Value.Of(el.Hi));
            arr.Add(Value.Of(o));
        }
        return Value.Of(arr);
    }

    public static Value CentersToValue(IReadOnlyList<Center> cs)
    {
        var arr = new List<Value>(cs.Count);
        foreach (var c in cs)
        {
            var o = new OrderedObject();
            o.Set("start", Value.Of(c.Start));
            o.Set("end", Value.Of(c.End));
            o.Set("ZD", Value.Of(c.ZD));
            o.Set("ZG", Value.Of(c.ZG));
            o.Set("n", Value.Of(c.N));
            arr.Add(Value.Of(o));
        }
        return Value.Of(arr);
    }

    /// <summary>
    /// Parse a list of centers (trend_type input). The fixtures may or may not
    /// carry start/end/n; defaults to 0 (the trend classifier only reads ZD/ZG).
    /// </summary>
    public static List<Center> CentersFromValue(Value v)
    {
        var arr = v.AsArray();
        var out_ = new List<Center>(arr.Count);
        foreach (var e in arr)
        {
            var o = e.AsObject();
            long start = o.TryGet("start", out var s) ? s.AsInt() : 0;
            long end = o.TryGet("end", out var en) ? en.AsInt() : 0;
            long n = o.TryGet("n", out var nv) ? nv.AsInt() : 0;
            out_.Add(new Center(
                start, end,
                o.Get("ZD").AsInt(),
                o.Get("ZG").AsInt(),
                n));
        }
        return out_;
    }
}
